pyccl/baryons.py

```diff
@@ -0,0 +1,137 @@
+        """
+        raise NotImplementedError("Do not use base `BaryonicEffects` class.")
+
+    @unlock_instance(mutate=True, argv=2)
+    def _new_pk(self, cosmo, pk2d, a_arr, lk_arr, pk_arr, logp, in_place):
+        # Returns a new Pk2D, either in place or from scratch.
+        if in_place:
+            lib.f2d_t_free(pk2d.psp)
+            status = 0
+            pk2d.psp, status = lib.set_pk2d_new_from_arrays(
+                lk_arr, a_arr, pk_arr.flatten(),
+                int(pk2d.extrap_order_lok),
+                int(pk2d.extrap_order_hik),
+                int(logp), status)
+            check(status, cosmo)
+            return None
+        else:
+            new = Pk2D(a_arr=a_arr, lk_arr=lk_arr, pk_arr=pk_arr,
+                       is_logp=logp,
+                       extrap_order_lok=pk2d.extrap_order_lok,
+                       extrap_order_hik=pk2d.extrap_order_hik)
+            return new
+
+
+class BaryonicEffectsBCM(BaryonicEffects):
+    """The BCM model boost factor for baryons.
+
+    .. note:: BCM stands for the "baryonic correction model" of Schneider &
+              Teyssier (2015; https://arxiv.org/abs/1510.06034). See the
+              `DESC Note <https://github.com/LSSTDESC/CCL/blob/master/doc\
+/0000-ccl_note/main.pdf>`_
+              for details.
+
+    .. note:: The boost factor is applied multiplicatively so that
+              :math:`P_{\\rm corrected}(k, a) = P(k, a)\\, f_{\\rm bcm}(k, a)`.
+
+    Args:
+        log10Mc (:obj:`float`): logarithmic mass scale of hot
+            gas suppression. Defaults to 14.08.
+        eta_b (:obj:`float`): ratio of escape to ejection radii (see
+            Teyssier et al. 2015). Defaults to 0.5.
+        k_s (:obj:`float`): Characteristic scale (wavenumber) of
+            the stellar component. Defaults to 55.0.
+    """
+    __repr_attrs__ = ("log10Mc", "eta_b", "k_s")
+
+    def __init__(self, log10Mc=14.08, eta_b=0.5, k_s=55.0):
+        self.log10Mc = log10Mc
+        self.eta_b = eta_b
+        self.k_s = k_s
+        self.apply_baryons = self._apply_boost
+
+    def boost_factor(self, cosmo, k, a):
+        """The BCM model boost factor for baryons.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): Cosmological parameters.
+            k (float or array_like): Wavenumber; Mpc^-1.
+            a (float): Scale factor.
+
+        Returns:
+            float or array_like: Correction factor to apply to
+                the power spectrum.
+        """
+        k_use = np.atleast_1d(k)
+
+        z = 1./a - 1.
+        kh = k_use / cosmo['h']
+        b0 = 0.105*self.log10Mc - 1.27
+        bfunc = b0 / (1. + (z/2.3)**2.5)
+        bfunc4 = (1-bfunc)**4
+        kg = 0.7 * bfunc4 * self.eta_b**(-1.6)
+        gf = bfunc / (1 + (kh/kg)**3) + 1. - bfunc
+        scomp = 1 + (kh / self.k_s)**2
+        fka = gf * scomp
+
+        if np.ndim(k) == 0:
+            fka = fka[0]
+        return fka
```

Comment:
In the first case, I'm fairly sure this will return a scalar.

Reply:
I don't think we should support this. Just like `numpy` and most other Python libraries we should add in the docs that "dimensions of size 1 are squeezed in the output" and have this as a rule everywhere throughout CCL. Either way, it is rare that users will pass an array consisting of a single number.

In fixing the docs, I am planning to include a `Returns` section to every docstring, which is extremely helpful for users.

Reply:
We do support it already in the halos module, and I do think it's good to be consistent. When you put this in a likelihood, sometimes it's hard to predict what size arrays you'll be passing around, and I very much prefer the output to be consistent with the input. I don't really care what numpy do.

And, by the way
```
>>> import numpy as np
>>> np.log10(np.array([1.0]))
array([0.])
```

Reply:
Okay, in case this behaviour is desired I am happy to leave dimensional checks as they are.
However, we can clean up a lot of code if we use a dispatcher that automatically does that, and not require that developers do it every time. I will create one and replace accordingly once we are done merging things to `v3_wip`.

Resolving.